pkg/syncer/upsync/upsync_process.go

```diff
@@ -0,0 +1,183 @@
+package upsync
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kcp-dev/kcp/pkg/syncer/shared"
+	. "github.com/kcp-dev/kcp/tmc/pkg/logging"
+	"github.com/kcp-dev/logicalcluster/v2"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+const (
+	ResourceVersionAnnotation = "kcp.dev/resource-version"
+)
+
+func (c *Controller) pruneUpstreamResource(ctx context.Context, gvr schema.GroupVersionResource, key string) error {
+	logger := klog.FromContext(ctx)
+	upstreamNamespace, upstreamName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logger.Error(err, "Invalid key")
+		return nil
+	}
+	err = c.upstreamClient.Cluster(c.syncTargetWorkspace).Resource(gvr).Namespace(upstreamNamespace).Delete(ctx, upstreamName, metav1.DeleteOptions{})
+	if err != nil {
+		logger.Error(err, "Could not remove resource upstream")
+		return nil
+	}
+	return nil
+}
+
+func (c *Controller) processDownstreamResource(ctx context.Context, gvr schema.GroupVersionResource, key string) error {
+	logger := klog.FromContext(ctx)
+	downstreamNamespace, downstreamName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logger.Error(err, "Invalid key")
+		return nil
+	}
+
+	logger = logger.WithValues(DownstreamNamespace, downstreamNamespace, DownstreamName, downstreamName)
+
+	var upstreamLocator *shared.NamespaceLocator
+	var locatorExists bool
+	var downstreamObject interface{}
+	// Namespace scoped resource
+	if downstreamNamespace != "" {
+		var nsObj runtime.Object
+		if nsObj, err = c.downstreamNamespaceLister.Get(downstreamNamespace); err != nil {
+			logger.Error(err, "Error getting downstream Namespace from downstream namespace lister", "ns", downstreamNamespace)
+		}
+		nsMeta, ok := nsObj.(metav1.Object)
+		if !ok {
+			logger.Info(fmt.Sprintf("Error: downstream ns expected to be metav1.Object got %T", nsObj))
+		}
+
+		upstreamLocator, locatorExists, err = shared.LocatorFromAnnotations(nsMeta.GetAnnotations())
+		if err != nil {
+			logger.Error(err, "Error ")
+		}
+
+		if !locatorExists || upstreamLocator == nil {
+			return nil
+		}
+		downstreamObject, err = c.downstreamClient.Resource(gvr).Namespace(downstreamNamespace).Get(ctx, downstreamName, metav1.GetOptions{})
+		if err != nil {
```

Comment:
You have to manage the case here and return.
Distinguish between NotFound and a real error.
- If the object is not found, we might want to delete the object upstream (we can do it since it has the same name and we know the upstream namespace through the locator)
- In case of other errors, just return the error.

Reply:
updated!